File: src/ng/canonicalize.py
"""JCS-profile canonicalization and digests for ztr-finding-1."""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
from typing import Any, Mapping


def canonicalize_json(value: Any) -> bytes:
    """RFC 8785-compatible profile: sorted keys, compact separators, UTF-8."""
    return json.dumps(
        value,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")


def canonicalize_envelope_for_signing(envelope: Mapping[str, Any]) -> bytes:
    """Envelope document used as HMAC message (signature field omitted)."""
    unsigned = {k: v for k, v in envelope.items() if k != "signature"}
    return canonicalize_json(unsigned)


def body_digest_hex(raw_body: bytes) -> str:
    return hashlib.sha256(raw_body).hexdigest()


def payload_digest_hex(*, plaintext: Mapping[str, Any] | None, ciphertext_b64: str | None) -> str:
    if plaintext is not None and ciphertext_b64 is not None:
        raise ValueError("exactly one payload mode")
    if plaintext is not None:
        payload_bytes = canonicalize_json(plaintext)
    elif ciphertext_b64 is not None:
        payload_bytes = base64.b64decode(ciphertext_b64, validate=True)
    else:
        raise ValueError("missing payload")
    return hashlib.sha256(payload_bytes).hexdigest()


def hmac_sha256_hex(secret: bytes, message: bytes) -> str:
    return hmac.new(secret, message, hashlib.sha256).hexdigest()


def verify_hmac_sha256_hex(secret: bytes, message: bytes, signature_hex: str) -> bool:
    expected = hmac_sha256_hex(secret, message)
    return hmac.compare_digest(expected, signature_hex.lower())

File: client/lib/ingest/canonicalize.dart
/// JCS-profile JSON canonicalization matching `src/canonicalize.py`.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

/// RFC 8785-compatible profile: sorted keys, compact separators, UTF-8.
Uint8List canonicalizeJson(Object? value) {
  return Uint8List.fromList(utf8.encode(_encode(value)));
}

String canonicalizeJsonString(Object? value) => _encode(value);

Uint8List canonicalizeEnvelopeForSigning(Map<String, Object?> envelope) {
  final unsigned = Map<String, Object?>.from(envelope)..remove('signature');
  return canonicalizeJson(unsigned);
}

String bodyDigestHex(Uint8List rawBody) => sha256.convert(rawBody).toString();

String payloadDigestHex({
  Map<String, Object?>? plaintext,
  String? ciphertextB64,
}) {
  if (plaintext != null && ciphertextB64 != null) {
    throw ArgumentError('exactly one payload mode');
  }
  late final Uint8List payloadBytes;
  if (plaintext != null) {
    payloadBytes = canonicalizeJson(plaintext);
  } else if (ciphertextB64 != null) {
    payloadBytes = Uint8List.fromList(base64Decode(ciphertextB64));
  } else {
    throw ArgumentError('missing payload');
  }
  return sha256.convert(payloadBytes).toString();
}

String hmacSha256Hex(Uint8List secret, Uint8List message) {
  final digest = Hmac(sha256, secret).convert(message);
  return digest.toString();
}

String _encode(Object? value) {
  if (value == null) return 'null';
  if (value is bool) return value ? 'true' : 'false';
  if (value is num) {
    if (value is double && (value.isNaN || value.isInfinite)) {
      throw ArgumentError('NaN/Infinity not allowed');
    }
    // Match Python json.dumps for ints/floats used in envelopes.
    if (value is int) return value.toString();
    final d = value.toDouble();
    if (d == d.truncateToDouble() && !d.isNegative) {
      // 1.0 → 1 in Python when it's an int-like? Actually Python keeps 1.0 as 1.0
      // for float. Only int type becomes without decimal.
    }
    var s = d.toString();
    if (s.endsWith('.0') && d == d.roundToDouble()) {
      // Dart toString on 1.0 is "1.0" — good, matches Python float.
    }
    return s;
  }
  if (value is String) return jsonEncode(value);
  if (value is List) {
    final parts = value.map(_encode).join(',');
    return '[$parts]';
  }
  if (value is Map) {
    final keys = value.keys.map((k) => k.toString()).toList()..sort();
    final parts = <String>[];
    for (final key in keys) {
      parts.add('${jsonEncode(key)}:${_encode(value[key])}');
    }
    return '{${parts.join(',')}}';
  }
  throw ArgumentError('unsupported JSON type: ${value.runtimeType}');
}
